gapic/src/main/com/google/gapid/widgets/ImagePanel.java

```diff
@@ -299,372 +313,451 @@ protected void onUiThreadError(Loadable.Message message) {
-    private static final int PREVIEW_WIDTH = 19; // Should be odd, so center pixel looks nice.
-    private static final int PREVIEW_HEIGHT = 11; // Should be odd, so center pixel looks nice.
-    private static final int PREVIEW_SIZE = 7;
-    private static final int MODE_TEXTURE = 0, MODE_CHECKER = 1, MODE_SOLID = 2;
-    private static final float[] CURSOR_DARK = new float[] { 0, 0, 0 };
-    private static final float[] CURSOR_LIGHT = new float[] { 1, 1, 1 };
-
-    private final GlComposite canvas;
-    private Image imageData = Image.EMPTY;
-    private final AtomicBoolean newImageData = new AtomicBoolean(false);
-    private double zoom = ZOOM_FIT;
-    private boolean flipped = false;
-    private float[] channels = new float[] { 1f, 1f, 1f, 1f };
-    private BackgroundMode backgroundMode = BackgroundMode.Checkerboard;
-    private Color backgroundColor;
-    private Pixel previewPixel = Pixel.OUT_OF_BOUNDS;
+    private static final VecD MIN_ZOOM_SIZE = new VecD(100, 100, 0);
 
-    private Shader shader;
-    private Buffer buffer;
-    private Texture texture;
-    private int[] uSize = new int[2], uOffset = new int[2];
-    private float[] uRange = new float[] { 0, 1 };
+    private final boolean naturallyFlipped;
+
+    private final ScrollBar scrollbars[];
+    private final ScenePanel canvas;
+    private final SceneData settings;
+    private Image image;
 
-    public ImageComponent(Composite parent) {
-      super(parent, SWT.BORDER | SWT.V_SCROLL | SWT.H_SCROLL);
+    private double scaleImageToViewMin = 0;
+    private double scaleImageToViewMax = Double.POSITIVE_INFINITY;
+    private double scaleImageToViewFit = 1;
+
+    private VecD viewSize = VecD.ZERO;
+    private VecD imageSize = VecD.ZERO;
+    private VecD imageOffset = VecD.ZERO;
+    private VecD imageOffsetMin = VecD.MIN;
+    private VecD imageOffsetMax = VecD.MAX;
+
+    private double scaleImageToView = 1.0;
+    private boolean zoomToFit;
+
+    public ImageComponent(Composite parent, boolean naturallyFlipped) {
+      super(parent, SWT.BORDER | SWT.V_SCROLL | SWT.H_SCROLL | SWT.NO_BACKGROUND);
       setLayout(new FillLayout(SWT.VERTICAL));
 
-      canvas = new GlComposite(this);
-      canvas.addListener(this);
+      this.naturallyFlipped = naturallyFlipped;
+
+      scrollbars = new ScrollBar[] { getHorizontalBar(), getVerticalBar() };
 
-      getHorizontalBar().addListener(SWT.Selection, e -> canvas.paint());
-      getVerticalBar().addListener(SWT.Selection, e -> canvas.paint());
+      settings = new SceneData();
+      settings.flipped = naturallyFlipped;
+      settings.borderWidth = BORDER_SIZE;
+      settings.borderColor = getDisplay().getSystemColor(SWT.COLOR_WIDGET_NORMAL_SHADOW);
+      settings.panelColor = getDisplay().getSystemColor(SWT.COLOR_WIDGET_BACKGROUND);
+      settings.backgroundMode = BackgroundMode.Checkerboard;
+      settings.checkerDark = new Color(getDisplay(), 192, 192, 192);
```

Comment:
Colors created via ```new Color``` need to be disposed. Best is to put them into the theme, so that this is done for you.